File: app/src/main/java/br/com/douglas/fukuhara/lodjinha/HomeActivity.java
package br.com.douglas.fukuhara.lodjinha;

import android.os.Bundle;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import br.com.douglas.fukuhara.lodjinha.view.AboutFragment;
import br.com.douglas.fukuhara.lodjinha.view.HomeFragment;

public class HomeActivity extends AppCompatActivity
        implements NavigationView.OnNavigationItemSelectedListener {

    private DrawerLayout mDrawerLayout;
    private TextView tvCustomToolbarTitle;

    private final int CURRENT_FRAGMENT = 0;
    private final String HOME_FRAGMENT = "home_fragment";
    private final String ABOUT_FRAGMENT = "about_fragment";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.home_activity);

        initNavDrawerComponents();
        initFragmentOnTop(savedInstanceState);
    }

    private void initNavDrawerComponents() {
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        mDrawerLayout = findViewById(R.id.home_drawer_layout);
        tvCustomToolbarTitle = findViewById(R.id.toolbar_custom_text);

        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, mDrawerLayout, toolbar, R.string.drawer_open_desc, R.string.drawer_close_desc);
        mDrawerLayout.addDrawerListener(toggle);
        toggle.syncState();

        NavigationView navigationView = findViewById(R.id.home_drawer_nav_view);
        navigationView.setItemIconTintList(null);
        navigationView.setNavigationItemSelectedListener(this);
    }

    private void initFragmentOnTop(Bundle savedInstanceState) {
        if(savedInstanceState == null) {
            loadFragment(new HomeFragment(), HOME_FRAGMENT);
        } else {
            handleToolbarCustomization(getSupportFragmentManager().getFragments().get(CURRENT_FRAGMENT));
        }
    }

    @Override
    public boolean onNavigationItemSelected(MenuItem menuItem) {
        int id = menuItem.getItemId();

        switch (id) {
            case R.id.home_drawer_menu_home:
                Fragment homeFragmentByTag = getSupportFragmentManager().findFragmentByTag(HOME_FRAGMENT);
                if (homeFragmentByTag == null || !homeFragmentByTag.isVisible()) {
                    loadFragment(new HomeFragment(), HOME_FRAGMENT);
                }
                break;
            case R.id.home_drawer_menu_about:
                Fragment aboutFragmentByTag = getSupportFragmentManager().findFragmentByTag(ABOUT_FRAGMENT);
                if (aboutFragmentByTag == null || !aboutFragmentByTag.isVisible()) {
                    loadFragment(new AboutFragment(), ABOUT_FRAGMENT);
                }
                break;
        }
        mDrawerLayout.closeDrawer(GravityCompat.START);
        return true;
    }

    @Override
    public void onBackPressed() {
        if (mDrawerLayout.isDrawerOpen(GravityCompat.START)) {
            mDrawerLayout.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    private void loadFragment(Fragment fragment, String tag) {
        handleToolbarCustomization(fragment);
        getSupportFragmentManager().beginTransaction().replace(R.id.fragment_container, fragment, tag).commit();
    }

    private void handleToolbarCustomization(Fragment fragment) {
        if (fragment instanceof AboutFragment) {
            getSupportActionBar().setDisplayShowTitleEnabled(true);
            getSupportActionBar().setTitle(R.string.about_toolbar_title);
            getSupportActionBar().setIcon(null);
            tvCustomToolbarTitle.setVisibility(View.GONE);
        } else if (fragment instanceof HomeFragment) {
            getSupportActionBar().setDisplayShowTitleEnabled(false);
            getSupportActionBar().setIcon(R.drawable.logo_navbar);
            tvCustomToolbarTitle.setVisibility(View.VISIBLE);
        }
    }
}

File: app/src/main/java/br/com/cemobile/lodjinha/ui/home/HomeActivity.kt
package br.com.cemobile.lodjinha.ui.home

import android.os.Bundle
import android.support.annotation.VisibleForTesting
import android.support.design.widget.NavigationView
import android.support.test.espresso.IdlingResource
import android.support.v4.app.FragmentTransaction
import android.support.v4.view.GravityCompat
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import br.com.cemobile.lodjinha.R
import br.com.cemobile.lodjinha.ui.about.AboutFragment
import br.com.cemobile.lodjinha.ui.base.BaseFragment
import br.com.cemobile.lodjinha.util.testing.EspressoIdlingResource
import kotlinx.android.synthetic.main.activity_home.*
import kotlinx.android.synthetic.main.app_bar.*

class HomeActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home)
        setupViews()
    }

    override fun onBackPressed() {
        if (drawerLayout.isDrawerOpen(GravityCompat.START)) {
            drawerLayout.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        R.id.action_settings -> true
        else -> super.onOptionsItemSelected(item)
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        when (item.itemId) {
            R.id.nav_home -> navigateToHome()
            R.id.nav_about -> navigateToAbout()
        }

        drawerLayout.closeDrawer(GravityCompat.START)
        return true
    }

    fun setTitle(title: String) {
        supportActionBar?.title = title
    }

    fun showToolbarLogo() {
        supportActionBar?.setLogo(R.drawable.logo_menu)
    }

    fun hideToolbarLogo() {
        supportActionBar?.setLogo(null)
    }

    private fun setupViews() {
        setSupportActionBar(toolbar)
        setupNavigationDrawer()

        if (supportFragmentManager.findFragmentById(R.id.contentLayout) == null) {
            supportFragmentManager.beginTransaction()
                .replace(R.id.contentLayout, HomeFragment.newInstance())
                .commit()
        }
    }

    private fun setupNavigationDrawer() {
        val toggle = ActionBarDrawerToggle(
                this, drawerLayout, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        drawerLayout.addDrawerListener(toggle)
        toggle.syncState()

        navigationView.setNavigationItemSelectedListener(this)
    }

    private fun setFragment(fragment: BaseFragment) {
        // does not replace current fragment if it is the same type
        val currentFragment = supportFragmentManager.findFragmentById(R.id.contentLayout) as BaseFragment
        if (currentFragment.getTagName() == fragment.getTagName()) {
            return
        }

        supportFragmentManager.beginTransaction()
            .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
            .replace(R.id.contentLayout, fragment)
            .addToBackStack(fragment.getTagName())
            .commit()
    }

    private fun navigateToHome() {
        setFragment(HomeFragment.newInstance())
    }

    private fun navigateToAbout() {
        setFragment(AboutFragment.newInstance())
    }

    @VisibleForTesting
    fun getCountingIdlingResource(): IdlingResource = EspressoIdlingResource.idlingResource

}
